aten/src/ATen/core/ivalue_inl.h

```diff
@@ -343,37 +343,29 @@ struct C10_EXPORT ivalue::Future final : c10::intrusive_ptr_target {
       finished_cv_.wait(lock);
     }
 
-    if (!eptr_) {
-      postWaitHook(value_);
-    }
+    synchronizeWithCurrentStreams();
   }
 
   /**
    * Wait on the future until it completes and throw an
    * exception if an error exists.
    */
   void waitAndThrow() {
-    std::unique_lock<std::mutex> lock(mutex_);
-    while (!completed_) {
-      finished_cv_.wait(lock);
-    }
+    wait();
 
     if (eptr_) {
       std::rethrow_exception(eptr_);
     }
-
-    postWaitHook(value_);
```

Comment:
One assumption needed to understand why this simplification works is that, when `eptr_ != nullptr` we have no events/dataptrs stored, hence `synchronizeWithCurrentStreams` is a no-op.